libtenzir/src/multi_series_builder.cpp

```diff
@@ -0,0 +1,571 @@
+      auto res = msb_->builder_raw_.parser_(s, nullptr);
+      auto& [value, diag] = res;
+      if (value) {
+        b.data(std::move(*value));
+      } else {
+        b.data(s);
+      }
+    },
+    [&](raw_pointer raw) {
+      raw->data_unparsed(std::string{s});
+    },
+  };
+  return std::visit(visitor, var_);
+}
+
+auto list_generator::data_unparsed(std::string s) -> void {
+  const auto visitor = detail::overload{
+    [&](tenzir::builder_ref b) {
+      auto res = msb_->builder_raw_.parser_(s, nullptr);
+      auto& [value, diag] = res;
+      if (value) {
+        b.data(std::move(*value));
+      } else {
+        b.data(s);
+      }
+    },
+    [&](raw_pointer raw) {
+      raw->data_unparsed(std::move(s));
+    },
+  };
+  return std::visit(visitor, var_);
+}
+
+auto list_generator::record() -> record_generator {
+  const auto visitor = detail::overload{
+    [&](tenzir::builder_ref b) {
+      return record_generator{msb_, b.record()};
+    },
+    [&](raw_pointer raw) {
+      return record_generator{msb_, raw->record()};
+    },
+  };
+  return std::visit(visitor, var_);
+}
+
+auto list_generator::list() -> list_generator {
+  const auto visitor = detail::overload{
+    [&](tenzir::builder_ref b) {
+      return list_generator{msb_, b.list()};
+    },
+    [&](raw_pointer raw) {
+      return list_generator{msb_, raw->list()};
+    },
+  };
+  return std::visit(visitor, var_);
+}
+
+auto series_to_table_slice(series array,
+                           std::string_view fallback_name) -> table_slice {
+  TENZIR_ASSERT(caf::holds_alternative<record_type>(array.type));
+  TENZIR_ASSERT(array.length() > 0);
+  if (array.type.name().empty()) {
+    array.type = tenzir::type{fallback_name, array.type};
+  }
+  auto* cast = dynamic_cast<arrow::StructArray*>(array.array.get());
+  TENZIR_ASSERT(cast);
+  auto arrow_schema = array.type.to_arrow_schema();
+  auto batch = arrow::RecordBatch::Make(std::move(arrow_schema), cast->length(),
+                                        cast->fields());
+  TENZIR_ASSERT(batch);
+  TENZIR_ASSERT_EXPENSIVE(batch->Validate().ok());
+  return table_slice{std::move(batch), std::move(array.type)};
+}
+auto series_to_table_slice(std::vector<series> data,
+                           std::string_view fallback_name)
+  -> std::vector<table_slice> {
+  auto result = std::vector<table_slice>{};
+  result.resize(data.size());
+  std::ranges::transform(
+    std::move(data), result.begin(), [fallback_name](auto& s) {
+      return series_to_table_slice(std::move(s), fallback_name);
+    });
+  return result;
+}
+} // namespace detail::multi_series_builder
+
+multi_series_builder::multi_series_builder(
+  policy_type policy, settings_type settings, diagnostic_handler& dh,
+  std::vector<type> schemas, data_builder::data_parsing_function parser)
+  : policy_{std::move(policy)},
+    settings_{std::move(settings)},
+    dh_{dh},
+    builder_raw_{std::move(parser), &dh, settings_.schema_only, settings_.raw} {
```

Comment:
Compare
~~~
builder_raw_{std::move(parser), &dh, settings_.schema_only, settings_.raw}
~~~
with
~~~
 data_builder(data_parsing_function parser
               = detail::data_builder::basic_parser,
               diagnostic_handler* dh = nullptr, bool schema_only = false,
               bool parse_schema_fields_only = false);
~~~
and the docs of `settings_type::raw`:
~~~
      // Whether to not parse fields that are not present in a known schema.
      bool raw = false;
~~~
What's going on here?

Reply:
I am not sure I understand the issue. Granted the double negation in the docs for `raw` is slightly confusing.

We made the decision that even with `--raw` fields that have specified type in the used schema should still be parsed.

Hence why the "internal" name in the `data_builder` was changed to `parse_schema_fields_only`, whereas the "external" name in the `multi_series_builder::settings_type`  was kept as `raw` in order to match the TQL argument name.

Reply:
Oh, `parse` does not refer to the parser itself here, but to the data parsing function used for strings? I was very confused by the name... but knowing that it refers to the function, it makes sense 🙈